tests/phpunit/integration/Core/Permissions/PermissionsTest.php

```diff
@@ -0,0 +1,206 @@
+		$this->assertFalse( $user->has_cap( Permissions::VIEW_DASHBOARD ) );
+		$this->assertFalse( $user->has_cap( Permissions::VIEW_MODULE_DETAILS ) );
+	}
+
+	public function test_authenticated_user_with_administrator_role_and_setup_complete() {
+		$user = self::factory()->user->create_and_get( array( 'role' => 'administrator' ) );
+		wp_set_current_user( $user->ID );
+		do_action( 'wp_login', $user->user_login, $user );
+
+		$context = new Context( GOOGLESITEKIT_PLUGIN_MAIN_FILE );
+		$auth    = new Authentication(
+			$context,
+			new Options( $context ),
+			new User_Options( $context, $user->ID )
+		);
+
+		$this->assertFalse( $auth->is_authenticated() );
+		$this->assertFalse( $auth->is_setup_completed() );
+		$this->assertFalse( $auth->verification()->has() );
+
+		// Setup the verification on the current user.
+		$auth->verification()->set( true );
+		// Fake a valid authentication token on the client.
+		$auth->get_oauth_client()->set_token(
+			array(
+				'access_token' => 'valid-auth-token',
+			)
+		);
+
+		// Mock a valid oauth token.
+		add_filter(
+			'googlesitekit_oauth_secret',
+			function () {
+				return array(
+					'web' => array(
+						'client_id'     => '1234567890-asdfasdfasdfasdfzxcvzxcvzxcvzxcv.apps.sitekit.withgoogle.com',
+						'client_secret' => 'x_xxxxxxxxxxxxxxxxxxxxxx',
+					),
+				);
+			}
+		);
+
+		// Override any existing filter to make sure the setup is marked as complete all the time.
+		add_filter( 'googlesitekit_setup_complete', '__return_true', 100 );
+
+		$this->assertTrue( $auth->is_authenticated() );
+		$this->assertTrue( $auth->is_setup_completed() );
+		$this->assertTrue( $auth->verification()->has() );
+
+		$this->get_instance( $auth )->register();
+
+		$this->assertTrue( has_filter( 'map_meta_cap' ) );
+		$this->assertTrue( has_filter( 'googlesitekit_user_data' ) );
+		$this->assertTrue( has_filter( 'user_has_cap' ) );
+
+		$this->assertTrue( $user->has_cap( Permissions::SETUP ) );
+		$this->assertTrue( $user->has_cap( Permissions::AUTHENTICATE ) );
+		$this->assertTrue( $user->has_cap( Permissions::VIEW_POSTS_INSIGHTS ) );
+		$this->assertTrue( $user->has_cap( Permissions::MANAGE_OPTIONS ) );
+		$this->assertTrue( $user->has_cap( Permissions::VIEW_DASHBOARD ) );
+		$this->assertTrue( $user->has_cap( Permissions::VIEW_MODULE_DETAILS ) );
+	}
+
+	public function test_get_capabilities() {
+		$capabilities = array(
+			Permissions::AUTHENTICATE        => true,
+			Permissions::SETUP               => true,
+			Permissions::VIEW_POSTS_INSIGHTS => true,
+			Permissions::VIEW_DASHBOARD      => true,
+			Permissions::VIEW_MODULE_DETAILS => true,
+			Permissions::MANAGE_OPTIONS      => true,
+		);
+
+		foreach ( Permissions::get_capabilities() as $capability ) {
+			$this->assertArrayHasKey( $capability, $capabilities );
+		}
+	}
+
+	protected function get_instance( Authentication $authentication = null ) {
+		return new Permissions(
```

Comment:
Nit: Most tests instantiate the main class in the test. This works fine, and makes the instance chainable off the method call but is a bit of a trade for readability. There are also cases where we need to instantiate `Context` differently in tests, which isn't needed here but worth noting.

I'm not strongly against using such methods for convenience, and there isn't a good reason not to use this, but if you'd prefer to keep it, let's name the method a little differently such as `new_instance` or `new_permissions` to make it a bit more clear what it's doing.

Reply:
Thanks, this is a good point I would remove this method and create the instance directly in each test method instead.